sched/addrenv/addrenv.c

```diff
@@ -237,62 +208,48 @@ int addrenv_switch(FAR struct tcb_s *tcb)
  *   Allocate an address environment for a new process.
  *
  * Input Parameters:
- *   tcb   - The tcb of the newly created task.
- *   ttype - The type of the task.
+ *   None
  *
  * Returned Value:
- *   This is a NuttX internal function so it follows the convention that
- *   0 (OK) is returned on success and a negated errno is returned on
- *   failure.
+ *   Pointer to the new address environment, or NULL if out of memory.
  *
  ****************************************************************************/
 
-int addrenv_allocate(FAR struct tcb_s *tcb, uint8_t ttype)
+FAR struct addrenv_s *addrenv_allocate(void)
 {
-  int ret = OK;
+  FAR struct addrenv_s *addrenv;
 
-  if ((ttype & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_KERNEL)
+  addrenv = (FAR struct addrenv_s *)kmm_zalloc(sizeof(struct addrenv_s));
+  if (addrenv)
     {
-      tcb->addrenv_own = NULL;
-    }
-  else
-    {
-      tcb->addrenv_own = (FAR struct addrenv_s *)
-        kmm_zalloc(sizeof(struct addrenv_s));
-      if (tcb->addrenv_own == NULL)
-        {
-          ret = -ENOMEM;
-        }
+      /* Take reference so this won't get freed */
+
+      addrenv->refs = 1;
     }
 
-  return ret;
+  return addrenv;
 }
 
 /****************************************************************************
  * Name: addrenv_free
  *
  * Description:
- *   Free an address environment for a process.
+ *   Unconditionally free an address environment.
```

Comment:
```suggestion
 *   Free an address environment.
```
Or remove `if (addrenv)` as those seems to go in conflict.

Reply:
Here "undonditionally" means "do not check for holders / references". The struct addrenv_s has a reference counter that protects it from being freed if someone is using it.

This function is intended to handle special cases where no one is for SURE not yet using it, e.g. inside elf_addrenv_alloc. This function is called to simply free the kmalloc'd memory obtained by addrenv_allocate().

The "if (addrenv)" test is simply a sanity check, so if someone gives a NULL pointer here the system won't crash.

Reply:
Ok. Let's keep it as is. But the system won't crash even if somebody gives a NULL pointer.

Reply:
I'll remove this function entirely, addrenv_drop() can do the same work.

Reply:
Yeah, the separated build will.. because of `DEBUGASSERT((mem == NULL) || kmm_heapmember(mem));`, but if `kmm_free` is mapped to `free`, then no issues

Reply:
I solved this by removing the function entirely, I realized after your comment it is not needed anyway :)

Reply:
Anyway `DEBUGASSERT((mem == NULL) || kmm_heapmember(mem));` and not `DEBUGASSERT((mem != NULL) && kmm_heapmember(mem));` seems to be odd to me from the first look.
Maybe we can think of unification of operation by
```
#  define kmm_free(p)            do \
               { \
                  DEBUGASSERT(((p) == NULL) || kmm_heapmember(p)); \
                  free(p);\
               } \
             while(0)
```

Reply:
Yeah... The `DEBUGASSERT((mem == NULL) || kmm_heapmember(mem));` does exactly either allow `NULL` or if not NULL the pointer should be a member of kernel heap.

Reply:
And I will take a look into `kmm_free` as `DEBUGASSERT((mem == NULL) || kmm_heapmember(mem));` seems to be not needed since `mm_free` contains `DEBUGASSERT(mm_heapmember(heap, mem));` inside